### PR TITLE
Hotfix/jm 7261 - General Performance Improvements

### PR DESCRIPTION
**Pull request by philhead1**
### JIRA link (if applicable) ###
https://centralgovernmentcgi.atlassian.net/browse/JM-7261

### Change description ###
* Optimised Active pools for bureau and also optimised summoning jurors

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java [hunk 1]
```diff
@@ -420,24 +421,28 @@ private List<JurorPool> getJurorPools(String login, String owner, PoolCreateRequ
...
             }
 
+            // Saving records (bulk)
+            jurorRepository.saveAll(jurorPools.stream().map(JurorPool::getJuror).toList());
+            jurorPoolRepository.saveAll(jurorPools);
+
+            // create a summons letter for juror
+            printDataService.bulkPrintSummonsLetter(jurorPools.stream()
+                .filter(jurorPool -> !Objects.equals(jurorPool.getStatus().getStatus(), IJurorStatus.DISQUALIFIED))
+                .toList());
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java:437]: it maybe worth creating an integration test here where we try and save an empty list... I have a feeling this may fail (But I could also be vastly incorrect)

Same for bulk print summons letter

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java [hunk 2]
```diff
@@ -501,7 +506,7 @@ private JurorPool createJurorPool(String login, String owner, Voters voter,
         juror.setResponded(false);
         juror.setContactPreference(null);
 
-        juror = jurorRepository.saveAndFlush(juror);
+        //juror = jurorRepository.save(juror);
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java:509]: Why do we no longer save the juror here?

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java [hunk 3]
```diff
@@ -513,7 +518,7 @@ private JurorPool createJurorPool(String login, String owner, Voters voter,
 
         jurorPool.setJuror(juror);
 
-        jurorPoolRepository.saveAndFlush(jurorPool);
+        //jurorPoolRepository.save(jurorPool);
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java:521]: Why do we no longer save the juror here?

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java (hunk 1 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java]: You were correct, but further testing the code it can never reach that point in it's current state. I've added a test to the create pool controller to verify that, also a bulk print summons letter tests for null list already exists.

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java (hunk 3 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java]: This is because the save is done later after we have created the new entity. Removed comment.

File: src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java (hunk 2 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/service/PoolCreateServiceImpl.java]: This is because the save is done later after we have created the new entity. Removed comment.

File: src/integration-test/java/uk/gov/hmcts/juror/api/bureau/controller/BureauOfficerAllocatedRepliesControllerTest.java [hunk 4]
```diff
@@ -107,7 +105,8 @@ public void bureauAllocationReplies_happy() throws Exception {
         assertThat(exchange.getStatusCode()).isEqualTo(HttpStatus.OK);
         assertThat(exchange.getBody().getBureauBacklogCount().getNonUrgent()).isEqualTo(4);
         assertThat(exchange.getBody().getBureauBacklogCount().getUrgent()).isEqualTo(3);
-        assertThat(exchange.getBody().getData().size()).isEqualTo(6);
```
Comment (Ben-Edwards-cgi) [inline src/integration-test/java/uk/gov/hmcts/juror/api/bureau/controller/BureauOfficerAllocatedRepliesControllerTest.java:110]: Why has this count changed from 6 to 2?

File: src/main/java/uk/gov/hmcts/juror/api/bureau/service/BureauOfficerAllocatedRepliesServiceImpl.java [hunk 5]
```diff
@@ -34,26 +29,31 @@ public BureauOfficerAllocatedResponses getBackLogData() {
...
             .findAll(UserQueries.activeBureauOfficers()));
 
+        Tuple backlogStats = jurorResponseRepository.getAssignRepliesStatistics();
+
         log.trace("Extracting backlog data");
         BureauBacklogCountData bureauBacklogCountDto = BureauBacklogCountData.builder()
-            .nonUrgent(bureauBacklogCountService.getBacklogNonUrgentCount())
-            .urgent(bureauBacklogCountService.getBacklogUrgentCount())
-            .allReplies(bureauBacklogCountService.getBacklogAllRepliesCount())
+            .nonUrgent(backlogStats.get(0, Integer.class).longValue())
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/bureau/service/BureauOfficerAllocatedRepliesServiceImpl.java:36]: Would be better to use the query dsl names instead of index here from a readability viewpoint

File: src/main/java/uk/gov/hmcts/juror/api/juror/domain/JurorResponseQueries.java [hunk 6]
```diff
@@ -144,8 +144,7 @@ public static BooleanExpression byAssignedUrgent(User staffMember) {
      * @return all responses assigned to staff
      */
     public static BooleanExpression byAssignedAll(User staffMember) {
-        return jurorResponse.staff.isNotNull()
-            .and(notClosed())
-            .and(assignedTo(staffMember));
+        return jurorResponse.staff.eq(staffMember)
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/juror/domain/JurorResponseQueries.java:147]: Would be better to use the assignedTo method (So if we change this logic one point of change)

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/IJurorDigitalResponseRepositoryMod.java [hunk 7]
```diff
@@ -0,0 +1,11 @@
+package uk.gov.hmcts.juror.api.moj.repository.jurorresponse;
+
+import com.querydsl.core.Tuple;
+
+import java.util.List;
+
+public interface IJurorDigitalResponseRepositoryMod {
+    Tuple getAssignRepliesStatistics();
+
+    List<Tuple> getAssignRepliesStatisticForUsers();
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/IJurorDigitalResponseRepositoryMod.java:10]: Returning a Tuple should be avoided if possible (Java objects make code more readable)

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+public class JurorDigitalResponseRepositoryModImpl implements IJurorDigitalResponseRepositoryMod {
+    @PersistenceContext
+    EntityManager entityManager;
+    private final QDigitalResponse digitalResponse = QDigitalResponse.digitalResponse;
+    private final QUser user = QUser.user;
+
+    @Override
+    public Tuple getAssignRepliesStatistics() {
+        JPAQuery<Tuple> query = getJpaQueryFactory().select(
+                new CaseBuilder().when(digitalResponse.staff.isNull()
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java:25]: I would assume we would also want to include paper replies here?

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java [hunk 9]
```diff
@@ -0,0 +1,63 @@
...
+                new CaseBuilder().when(
+                    digitalResponse.processingStatus.eq(ProcessingStatus.TODO)
+                        .and(digitalResponse.urgent.isTrue().or(digitalResponse.superUrgent.isTrue()))
+                ).then(1).otherwise(0).sum(),
+                new CaseBuilder().when(
+                    digitalResponse.processingStatus.eq(ProcessingStatus.TODO)
+                ).then(1).otherwise(0).sum()
+            ).from(digitalResponse)
+            .join(user)
+            .on(user.eq(digitalResponse.staff).and(user.owner.eq(SecurityUtil.BUREAU_OWNER)))
```
Comment (Ben-Edwards-cgi) [inline src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java:55]: Can not use user.owner (This is being removed) should use User.userType = UserType.BUREAU

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java (hunk 8 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java:25]: Checked this with liam earlier and this is for digital only

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java (hunk 9 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/JurorDigitalResponseRepositoryModImpl.java:55]: Changed

File: src/main/java/uk/gov/hmcts/juror/api/bureau/service/BureauOfficerAllocatedRepliesServiceImpl.java (hunk 5 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/bureau/service/BureauOfficerAllocatedRepliesServiceImpl.java:36]: Updated 😄

File: src/main/java/uk/gov/hmcts/juror/api/juror/domain/JurorResponseQueries.java (hunk 6 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/juror/domain/JurorResponseQueries.java:147]: Updated 😄

File: src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/IJurorDigitalResponseRepositoryMod.java (hunk 7 above)
Comment (philhead1) [reply src/main/java/uk/gov/hmcts/juror/api/moj/repository/jurorresponse/IJurorDigitalResponseRepositoryMod.java:10]: Would normally avoid but in this case as the query is generic it made sense to use a Tuple

File: src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/CreatePoolControllerITest.java [hunk 10]
```diff
@@ -1530,6 +1530,30 @@ public void addCitizensToCoronerPool_courtUser() throws Exception {
...
+        PoolCreateRequestDto poolCreateRequest = setUpPoolCreateRequestDto();
+        poolCreateRequest.setCitizensToSummon(1);
+
+        final URI uri = URI.create("/api/v1/moj/pool-create/create-pool");
+
+        httpHeaders.set(HttpHeaders.AUTHORIZATION, bureauJwt);
+        RequestEntity<PoolCreateRequestDto> requestEntity = new RequestEntity<>(poolCreateRequest, httpHeaders,
+            HttpMethod.POST, uri);
+        ResponseEntity<String> response = template.exchange(requestEntity, String.class);
+        assertThat(response.getStatusCode()).isEqualTo(HttpStatus.INTERNAL_SERVER_ERROR);
```
Comment (Ben-Edwards-cgi) [inline src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/CreatePoolControllerITest.java:1554]: I would expect a BVR here instead of an INTERNAL_SERVER_ERROR

**Review by Ben-Edwards-cgi — CHANGES_REQUESTED** [review]

File: src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/CreatePoolControllerITest.java (hunk 10 above)
Comment (philhead1) [reply src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/CreatePoolControllerITest.java:1554]: That's existing logic of the create pool controller that it throws a runtime exception, would update but would impact front-end as it will need changing to handle the new error (in context to this endpoint).

Comment (philhead1) [reply src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/CreatePoolControllerITest.java]: Just updated this from the discussion earlier this morning to use BVR

Comment (Ben-Edwards-cgi) [reply src/integration-test/java/uk/gov/hmcts/juror/api/moj/controller/CreatePoolControllerITest.java:1554]: Sorry this is my miss understanding this should be an INTERNAL_SERVER_ERROR as it is chucked inside a try/catch block on any unexpected exception.

(I thought it was business logic checking from the test name)

**Review by Ben-Edwards-cgi — APPROVED** [review]
